pkg/controller/utils/utils.go

```diff
@@ -1065,22 +1065,83 @@ func MaintainInstallationFinalizer(
 
 		// Check if the namespaced secondaryResources are still present.
 		// Keep track of all the secondary resources that the main resource creates.
-		// Only delete the finalizer if all of the secondary resources are deleted.
+		// Only delete the finalizer if all of the secondary resources are deleted and there are no lingering Pods.
 		for _, secondaryResource := range secondaryResources {
 			err := c.Get(ctx, types.NamespacedName{Namespace: secondaryResource.GetNamespace(), Name: secondaryResource.GetName()}, secondaryResource)
 			if err != nil && !errors.IsNotFound(err) {
 				return err
 			} else if errors.IsNotFound(err) {
-				log.Info("Object no longer exists.", "object", secondaryResource)
-			} else {
-				log.Info("Object is still present, waiting for termination", "object", secondaryResource)
+				log.V(2).Info("Object no longer exists.", "object", secondaryResource)
+			} else if err == nil {
+				log.V(2).Info("Object is still present, waiting for termination", "object", secondaryResource)
 				return nil
 			}
+
+			// If the secondary resource itself is gone, ensure there are no Pods left over that would still rely on CNI.
+			// Only applicable for workload types that create Pods.
+			switch obj := secondaryResource.(type) {
+			case *appsv1.Deployment:
+				matchLabels := getWorkloadMatchLabels(obj)
+				return checkLingeringPodsForWorkload(ctx, c, obj, matchLabels, "Deployment", "deployment")
+			case *appsv1.DaemonSet:
+				matchLabels := getWorkloadMatchLabels(obj)
+				return checkLingeringPodsForWorkload(ctx, c, obj, matchLabels, "DaemonSet", "daemonset")
+			case *appsv1.StatefulSet:
+				matchLabels := getWorkloadMatchLabels(obj)
+				return checkLingeringPodsForWorkload(ctx, c, obj, matchLabels, "StatefulSet", "statefulset")
+			default:
+				// Non-workload resource; no Pod listing needed.
+			}
 		}
-		log.Info("All objects no longer exist. Removing finalizer", "finalizer", finalizer)
+		log.V(2).Info("All objects and their Pods no longer exist. Removing finalizer", "finalizer", finalizer)
 		RemoveInstallationFinalizer(installation, finalizer)
 	}
 
 	// Update the installation with any finalizer changes.
 	return c.Patch(ctx, installation, patchFrom)
 }
+
+func hasLingeringPods(ctx context.Context, c client.Client, namespace string, matchLabels map[string]string) (bool, int, error) {
+	if matchLabels == nil {
+		return false, 0, nil
+	}
+	podList := &corev1.PodList{}
+	if err := c.List(
+		ctx,
+		podList,
+		client.InNamespace(namespace),
+		client.MatchingLabels(matchLabels),
+	); err != nil {
+		return false, 0, err
+	}
+	count := len(podList.Items)
+	return count > 0, count, nil
+}
+
+func getWorkloadMatchLabels(obj client.Object) map[string]string {
```

Comment:
We could simplify the logic in `MaintainInstallationFinalizer` if we moved the select statement inside of this function. Then, MaintainInstallationFinalizer would look more like this:

```go
sel, ok := getWorkloadMatchLabels(secondaryResource)
if ok {
  checkLingeringPodsForWorkload(ctx, c, secondaryResource, sel)
}
```

Or even just 

```go
  checkLingeringPodsForWorkload(ctx, c, secondaryResource, sel)
```

And have `checkLingeringPodsForWorkload` responsible for returning `nil` if not an applicable type, and parsing the selector if it is.